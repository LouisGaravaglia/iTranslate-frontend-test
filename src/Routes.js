import React from 'react';
import {Redirect, Route, Switch} from 'react-router-dom';
import Search from './Search';
import Browse from "./Browse";
import BrowseByArtists from "./BrowseByArtists";
import BrowseByGenre from "./BrowseByGenre";
import BrowseByDanceability from "./BrowseByDanceability";

const Routes = () => {

  return (
    <div className="main">
      <Switch>
        <Route exact path="/"><Search/></Route>
        <Route exact path="/browse"><Browse/></Route>
        <Route exact path="/browse/artists"><BrowseByArtists/></Route>
        <Route exact path="/browse/genres"><BrowseByGenre/></Route>
        <Route exact path="/browse/danceability"><BrowseByDanceability/></Route>
        <Redirect to="/"/>
      </Switch>
    </div>
  );
};

export default Routes;