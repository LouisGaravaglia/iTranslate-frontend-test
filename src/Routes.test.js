import React from 'react'
import {render, cleanup} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import {Provider} from 'react-redux'
import {createStore} from "redux";
import {createMemoryHistory} from 'history'
import {Router} from 'react-router-dom'
import Routes from "./Routes";

//CREATE A MOCK INTERSECTION OBSERVER CLASS FOR TESTING AND APPEND IT TO THE WINDOW OBJECT
const mockIntersectionObserver = class {
  constructor() {}
  observe() {}
  unobserve() {}
  disconnect() {}
};

beforeEach(() => {
  window.IntersectionObserver = mockIntersectionObserver;
});

afterEach(cleanup);

const startingState = {
  selectedTrack: {trackName: "chanel", trackId: 42}, 
  errors: {translationError: false, languageError: false, lyricsError: false, searchError: false, generalError: false}
};

/**
 * reducer function to mimic the reducer used in Redux.
 * @param {object} state - holds an object of data used within the components
 * @param {object} action - object where the values are strings used to fire certain actions
 */
function reducer(state = startingState, action) {
  switch (action.type) {
    case "ADD_TRACKS":
      return {...state, tracks: action.tracks};
    default:
      return state;
  };
};

/**
 * Function that renders the component within a redux environment since most 
 * components rely on redux for data in order to mount.
 * @param {ReactComponent} component - private spotify client id
 * @param {function} store - creates the react store using the createStore method imported from redux
 */
function renderWithRedux(
  component,
  {initialState, store = createStore(reducer, initialState)} = {}
) {
  return {
    ...render(<Provider store={store}>{component}</Provider>)
  };
};

//SMOKE TEST
describe('Smoke Test for Routes component', () => {

  it('renders without crashing', () => {
    const history = createMemoryHistory();

    renderWithRedux(
      <Router history={history}>
        <Routes />
      </Router>
    );
  });
});

//SNAPSHOT TEST
describe('Snapshot Test for Routes component', () => {

  it('matches snapshot', () => {
    const history = createMemoryHistory();

    const {asFragment} = renderWithRedux(
      <Router history={history}>
        <Routes />
      </Router>);
    expect(asFragment()).toMatchSnapshot();
  });
});