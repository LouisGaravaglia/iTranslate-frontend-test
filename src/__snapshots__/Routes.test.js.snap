// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test for Routes component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="main"
  >
    <div
      class="Flash-Messages-Container"
    />
    <div
      style="background-color: rgb(29, 77, 190);"
    >
      <div
        style="opacity: 0;"
      >
        <div
          class="Search-Field"
        >
          <div
            class="Search-Field-Content"
          >
            <h1
              class="Search-Landing"
            >
              Find your song!
            </h1>
            <form
              class="Search-Input-Container"
            >
              <div>
                <input
                  class="Search-Landing-Text"
                  id="SearchVal"
                  name="searchVal"
                  type="text"
                  value=""
                />
                <button
                  type="submit"
                >
                  <span
                    class="Hover-Box"
                    style="display: inline-block; transform: scale(1);"
                  >
                    <i
                      class="fa fa-search icon Search-Landing-Button"
                    />
                  </span>
                </button>
              </div>
            </form>
          </div>
          <div
            class="Search-Field-Filler"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
